libraries/Wire/src/Wire.cpp

```diff
@@ -132,6 +135,81 @@ bool TwoWire::setPins(int sdaPin, int sclPin)
     return !i2cIsInit(num);
 }
 
+bool TwoWire::allocateWireBuffer(void)
+{
+    // or both buffer can be allocated or none will be
+    if (rxBuffer == NULL) {
+            rxBuffer = (uint8_t *)malloc(BufferSize);
+            if (rxBuffer == NULL) {
+                log_e("Can't allocate memory for I2C_%d rxBuffer", num);
+                return false;
+            }
+    }
+    if (txBuffer == NULL) {
+            txBuffer = (uint8_t *)malloc(BufferSize);
+            if (txBuffer == NULL) {
+                log_e("Can't allocate memory for I2C_%d txBuffer", num);
+                freeWireBuffer();  // free rxBuffer for safety!
+                return false;
+            }
+    }
+    // in case both were allocated before, they must have the same size. All good.
+    return true;
+}
+
+void TwoWire::freeWireBuffer(void)
+{
+    if (rxBuffer != NULL) {
+        free(rxBuffer);
+        rxBuffer = NULL;
+    }
+    if (txBuffer != NULL) {
+        free(txBuffer);
+        txBuffer = NULL;
+    }
+}
+
+size_t TwoWire::setBufferSize(size_t bSize)
+{
+    // Maximum size .... HEAP limited ;-)
+    if (bSize < 32) {    // 32 bytes is the I2C FIFO Len for ESP32/S2/S3/C3
+        log_e("Minimum Wire Buffer size is 32 bytes");
+        return 0;
+    }
+
+#if !CONFIG_DISABLE_HAL_LOCKS
+    if(lock != NULL){
+        //acquire lock
+        if(xSemaphoreTake(lock, portMAX_DELAY) != pdTRUE){
+            log_e("could not acquire lock");
+            return 0;
+        }
+#endif
```

Comment:
If you call `Wire.setBufferSize()` before calling `Wire.begin()`, nothing will happen, because the `lock` is not created yet. WE need to be the same as it is in `TwoWire::begin()` function.

```
#if !CONFIG_DISABLE_HAL_LOCKS
    if(lock == NULL){
        lock = xSemaphoreCreateMutex();
        if(lock == NULL){
            log_e("xSemaphoreCreateMutex failed");
            return false;
        }
    }
    //acquire lock
    if(xSemaphoreTake(lock, portMAX_DELAY) != pdTRUE){
        log_e("could not acquire lock");
        return false;
    }
```

Reply:
In this case, we just must set bufferSize (I changed the captilization for keeping style) before returning, because `begin()` will `malloc()` the memory.

Reply:
I added a `else` for the `if(lock != NULL) { ...} else { bufferSize = bSize; }` to solve it.

Reply:
Why not to create the lock like its in `begin()` or `setPin()`? Then there is no need to `bufferSize = bSize` twice so the code will be simpler :)

```
size_t TwoWire::setBufferSize(size_t bSize)
{
    // Maximum size .... HEAP limited ;-)
    if (bSize < 32) {    // 32 bytes is the I2C FIFO Len for ESP32/S2/S3/C3
        log_e("Minimum Wire Buffer size is 32 bytes");
        return 0;
    }

#if !CONFIG_DISABLE_HAL_LOCKS
    if(lock == NULL){
        lock = xSemaphoreCreateMutex();
        if(lock == NULL){
            log_e("xSemaphoreCreateMutex failed");
            return false;
        }
        //acquire lock
        if(xSemaphoreTake(lock, portMAX_DELAY) != pdTRUE){
            log_e("could not acquire lock");
            return 0;
        }
    }
#endif
    // allocateWireBuffer allocates memory for both pointers or just free them
    if (rxBuffer != NULL || txBuffer != NULL) {
        // if begin() has been already executed, memory size changes... data may be lost. We don't care! :^)
        if (bSize != BufferSize) {
            // we want a new buffer size ... just reset buffer pointers and allocate new ones
            freeWireBuffer();
            BufferSize = bSize;
            if (!allocateWireBuffer()) {
                // failed! Error message already issued
                bSize = 0; // returns error
                log_e("Buffer allocation failed");
            }
            log_d("buff size = %d", BufferSize);
        } // else nothing changes, all set!
    } else {
        // no memory allocated yet, just change the size value - allocation in begin()
        BufferSize = bSize;
        log_d("buff size = %d", BufferSize);
    }
#if !CONFIG_DISABLE_HAL_LOCKS
    //release lock
    xSemaphoreGive(lock);
    
#endif
    return bSize;
}
```

Reply:
Adding the extra lock code is more expensive in terms of flash consumption, but, that's ok.  
Done.

Reply:
Opps... there is an error in the lock code... if it is not NULL, it is never acquired.... fixing it.